sources/address.c

```diff
@@ -0,0 +1,281 @@
+ */
+static bool od_address_check_hostname(struct sockaddr_storage *client_sa, const char *hostname)
+{
+	struct addrinfo *gai_result, *gai;
+	int ret;
+	bool found;
+
+	char client_hostname[NI_MAXHOST];
+
+	ret = getnameinfo(client_sa, sizeof(*client_sa),
+			  client_hostname, sizeof(client_hostname),
+			  NULL, 0,
+			  NI_NAMEREQD);
+
+	if (ret != 0)
+		return false;
+
+	/* Now see if remote host name matches this pg_hba line */
+	if (!od_address_hostname_match(hostname, client_hostname))
+		return false;
+
+	/* Lookup IP from host name and check against original IP */
+	ret = getaddrinfo(client_hostname, NULL, NULL, &gai_result);
+	if (ret != 0)
+		return false;
+
+	found = false;
+	for (gai = gai_result; gai; gai = gai->ai_next)
+	{
+		if (gai->ai_addr->sa_family == client_sa->ss_family)
+		{
+			if (gai->ai_addr->sa_family == AF_INET)
+			{
+				if (od_address_ipv4eq((struct sockaddr_in *)gai->ai_addr,
+						      (struct sockaddr_in *)client_sa))
+				{
+					found = true;
+					break;
+				}
+			}
+			else if (gai->ai_addr->sa_family == AF_INET6)
+			{
+				if (od_address_ipv6eq((struct sockaddr_in6 *)gai->ai_addr,
+						      (struct sockaddr_in6 *)&client_sa))
+				{
+					found = true;
+					break;
+				}
+			}
+		}
+	}
+
+	if (gai_result)
+		freeaddrinfo(gai_result);
+
+	return found;
+}
+
+bool od_address_validate(od_address_range_t *address_range, struct sockaddr_storage *sa)
+{
+	if (address_range->is_hostname)
+		return od_address_check_hostname(sa, address_range->string_value);
+
+	if (address_range->addr.ss_family != sa->ss_family)
+		return false;
+
+	if (sa->ss_family == AF_INET) {
+		struct sockaddr_in *sin = (struct sockaddr_in *)sa;
+		struct sockaddr_in *addr = (struct sockaddr_in *)&address_range->addr;
+		struct sockaddr_in *mask = (struct sockaddr_in *)&address_range->mask;
+		in_addr_t client_addr = sin->sin_addr.s_addr;
+		in_addr_t client_net = mask->sin_addr.s_addr & client_addr;
+		return (client_net ^ addr->sin_addr.s_addr) == 0;
+	} else if (sa->ss_family == AF_INET6) {
+		struct sockaddr_in6 *sin = (struct sockaddr_in6 *)sa;
+		struct sockaddr_in6 *addr = (struct sockaddr_in6 *)&address_range->addr;
+		struct sockaddr_in6 *mask = (struct sockaddr_in6 *)&address_range->mask;
+		for (int i = 0; i < 16; ++i) {
+			uint8_t client_net_byte = mask->sin6_addr.s6_addr[i] &
+						  sin->sin6_addr.s6_addr[i];
+			if (client_net_byte ^ addr->sin6_addr.s6_addr[i]) {
+				return false;
+			}
+		}
+		return true;
+	}
+
+	return false;
+}
+
+int od_address_hostname_validate(char *hostname)
+{
+	regex_t regex;
+	char *valid_rfc952_hostname_regex = "^(\\.?(([a-zA-Z]|[a-zA-Z][a-zA-Z0-9\\-]*[a-zA-Z0-9])\\.)*([A-Za-z]|[A-Za-z][A-Za-z0-9\\-]*[A-Za-z0-9]))$";
+	int reti = regcomp(&regex, valid_rfc952_hostname_regex, REG_EXTENDED);
+	if (reti)
+		return -1;
+	reti = regexec(&regex, hostname, 0, NULL, 0);
+	if (reti == 0)
+		return 0;
+	else
+		return 1;
+}
+
+uint32 od_address_bswap32(uint32 x)
```

Comment:
BTW we have od_hba_bswap32().
But I'd prefer to have od_bswap32() once and for all :)

Reply:
thanks, done